### PR TITLE
Feature/47 resume details

### PR DESCRIPTION
**Pull request by alejandro-castellon**
[AB#47](https://dev.azure.com/funjaladev/7867a1c5-5ac7-4c7f-baff-87d740bf4fc1/_workitems/edit/47)

### CODE REVIEW

File: src/components/Resume/ResumeDetail/ResumeDetail.jsx [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+      .then((result) => {
+        userRes = result;
+      })
+      .then(() => setUser(userRes));
+  };
+  useEffect(refresh, [userId]);
+
+  const [readOnly, setReadOnly] = useState(true);
+  const [showEdit, setShowEdit] = useState("Edit");
+  const classes = useStyles();
```
Comment (jorge-lema-jalafund) [inline src/components/Resume/ResumeDetail/ResumeDetail.jsx]: Please, move these statements  below **line 23** to keep some order in the code.
  ``` const [user, setUser] = useState(); ```

File: src/index.js [hunk 2]
```diff
@@ -1,11 +1,12 @@
...
 import ReactDOM from "react-dom";
 import "./index.css";
-import App from "./components/App/App";
+//import App from "./components/App/App";
+import Resume from "./components/Resume/ResumeDetail/ResumeDetail";
 
 ReactDOM.render(
   <React.StrictMode>
-    <App />
+    <Resume userId="1" />
```
Comment (jorge-lema-jalafund) [inline src/index.js]: we need to render the app component here. Don't worry about your component not being available when the webapp is running, as long as you have tested it and make sure there's no bugs then you don't have to worry. Once your teammates have completed their tasks you will be able to attach your component where it belongs.

File: src/index.js [hunk 3]
```diff
@@ -1,11 +1,12 @@
 import React from "react";
 import ReactDOM from "react-dom";
 import "./index.css";
-import App from "./components/App/App";
+//import App from "./components/App/App";
+import Resume from "./components/Resume/ResumeDetail/ResumeDetail";
```
Comment (jorge-lema-jalafund) [inline src/index.js]: we don't need to import Resume component here

File: src/components/Resume/ResumeDetail/ResumeDetail.jsx [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+  const edit = (event) => {
+    event.preventDefault();
+    setReadOnly(!readOnly);
+    if (readOnly) {
+      setShowEdit("Save");
+      alert("You can edit");
+    } else {
+      setShowEdit("Edit");
+      alert("Saved");
```
Comment (jorge-lema-jalafund) [inline src/components/Resume/ResumeDetail/ResumeDetail.jsx]: these alerts are good for now. They don't provide a good user experience so **we are removing them as soon as possible**.
Please leave some comments in your code about why we are using those alerts right now and that they are going to be replaced.

File: src/components/Resume/ResumeDetail/ResumeApi.js [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+const instance = axios.create({
+  baseURL: "https://jsonplaceholder.typicode.com/users/",
+});
+
+export async function GetUser(user) {
+  try {
+    let result = await (await instance.get(user)).data;
+    return result;
+  } catch (error) {
+    console.log(error);
```
Comment (jorge-lema-jalafund) [inline src/components/Resume/ResumeDetail/ResumeApi.js]: remove this console.log and any other that you find in your code please.
Those are only used for developers, we don't want to provide that information to final users.
Handle the error, you have to decide what you are going to do or return if you get an error while fetching data.

File: src/components/Resume/ResumeDetail/ResumeDetail.jsx [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    "& > *": {
+      margin: theme.spacing(1),
+    },
+  },
+}));
+
+function Resume({ userId }) {
+  const [user, setUser] = useState();
+
+  const refresh = () => {
```
Comment (jorge-lema-jalafund) [inline src/components/Resume/ResumeDetail/ResumeDetail.jsx]: refresh function is resolving a promise coming from your api call. You need to handle the error and possible results coming from your ResumeApi -> GetUser()

**Review by jorge-lema-jalafund — CHANGES_REQUESTED** [review]
Solve comments

File: src/index.js [hunk 7]
```diff
@@ -1,11 +1,12 @@
 import React from "react";
 import ReactDOM from "react-dom";
 import "./index.css";
-import App from "./components/App/App";
+//import App from "./components/App/App";
```
Comment (martin-callejas-jalafund) [inline src/index.js]: :eyes:

**Review by freddy-perez-jalafund — DISMISSED** [review]

File: package.json [hunk 8]
```diff
@@ -4,9 +4,11 @@
   "private": true,
   "dependencies": {
     "@material-ui/core": "^4.12.1",
+    "@material-ui/lab": "^4.0.0-alpha.60",
```
Comment (DaleneyRuben) [inline package.json]: What is it for?

File: src/components/Resume/ResumeDetail/ResumeApi.js [hunk 9]
```diff
@@ -0,0 +1,14 @@
+import axios from "axios";
+
+const instance = axios.create({
+  baseURL: "https://jsonplaceholder.typicode.com/users/",
+});
+
+export async function GetUser(user) {
+  try {
+    let result = await (await instance.get(user)).data;
```
Comment (DaleneyRuben) [inline src/components/Resume/ResumeDetail/ResumeApi.js]: why the double await?
I'd think that the following works: 
```
const result = await instance.get(user).data
```

File: src/components/Resume/ResumeDetail/ResumeApi.js [hunk 10]
```diff
@@ -0,0 +1,14 @@
+import axios from "axios";
+
+const instance = axios.create({
+  baseURL: "https://jsonplaceholder.typicode.com/users/",
+});
+
+export async function GetUser(user) {
```
Comment (DaleneyRuben) [inline src/components/Resume/ResumeDetail/ResumeApi.js]: Only functional components start with uppercase, in this case **getUser** is the right function name

File: src/components/Resume/ResumeDetail/ResumeDetail.jsx [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+    let userRes;
+    GetUser(userId)
+      .then((result) => {
+        userRes = result;
+      })
+      .then(() => setUser(userRes));
```
Comment (DaleneyRuben) [inline src/components/Resume/ResumeDetail/ResumeDetail.jsx]: could you try changing to use async await please? It's better to async/await over promises

File: src/components/Resume/ResumeDetail/ResumeDetail.jsx [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+  };
+  useEffect(refresh, [userId]);
```
Comment (DaleneyRuben) [inline src/components/Resume/ResumeDetail/ResumeDetail.jsx]: NABD but you're not reusing the refresh function so you could move all that code here

File: src/components/Resume/ResumeDetail/ResumeDetail.jsx [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+    setReadOnly(!readOnly);
+    if (readOnly) {
+      setShowEdit("Save");
+      alert("You can edit");
+    } else {
+      setShowEdit("Edit");
+      alert("Saved");
+    }
+  };
+  return user ? (
```
Comment (DaleneyRuben) [inline src/components/Resume/ResumeDetail/ResumeDetail.jsx]: you could return early here

```
if (!user) return // or return null

return  <Grid container className={classes.root} spacing={5}>
      <Grid item xs={12}>
        <Typography variant="h4" color="primary">
          Resume Details
        </Typography>
....
```

File: src/index.js (hunk 7 above)
Comment (DaleneyRuben) [reply src/index.js]: Why did you comment the app component? @alejandro-castellon-jalafund

**Review by DaleneyRuben — CHANGES_REQUESTED** [review]

File: src/components/Resume/ResumeDetail/ResumeDetail.jsx [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+        <TextField
+          label="Biography"
+          defaultValue={user.website}
+          InputProps={{
+            readOnly: readOnly,
+          }}
+          fullWidth
+          variant="outlined"
+          multiline
+          rows={4}
```
Comment (DaleneyRuben) [inline src/components/Resume/ResumeDetail/ResumeDetail.jsx]: can you order all this props alphabetically please? here and in all components you're passing props

**Review by DaleneyRuben — CHANGES_REQUESTED** [review]

File: src/index.js (hunk 7 above)
Comment (alejandro-castellon) [reply src/index.js]: Just to try my code, already fixed

File: package.json (hunk 8 above)
Comment (alejandro-castellon) [reply package.json]: There are some components like "Alert" that need this dependencie

File: src/components/Resume/ResumeDetail/ResumeApi.js (hunk 9 above)
Comment (alejandro-castellon) [reply src/components/Resume/ResumeDetail/ResumeApi.js]: `const result = await instance.get(user).data`
Receives an undefined object, meanwhile the double await returns the real data

File: package.json (hunk 8 above)
Comment (DaleneyRuben) [reply package.json]: We don't need alerts, please remove this dependency

Comment (DaleneyRuben) [reply package.json:7]: We don't need alerts, please remove this dependency

File: src/components/Resume/ResumeDetail/ResumeApi.js [hunk 15]
```diff
@@ -0,0 +1,26 @@
+import { useEffect, useState } from "react";
```
Comment (DaleneyRuben) [inline src/components/Resume/ResumeDetail/ResumeApi.js]: you should change the filename to useFetch,js

File: src/components/Resume/ResumeDetail/ResumeApi.js [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+import axios from "axios";
+
+function useFetch(url) {
+  const [data, setData] = useState();
+  const [error, setError] = useState();
+
+  useEffect(() => {
+    async function fetchData() {
+      try {
+        const res = await (await axios.get(url)).data;
```
Comment (DaleneyRuben) [inline src/components/Resume/ResumeDetail/ResumeApi.js]: why the double await? I'm pretty sure the following works:
```
const { data } = await axios.get(url)
setData(data)
```

before you used an axios intance I think that's why you had to use double await but in this case it's not necessary

File: src/components/Resume/ResumeDetail/ResumeDetail.jsx [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+  margin: {
+    "& > *": {
+      margin: theme.spacing(1),
+    },
+  },
+  root: {
+    width: 500,
+  },
+}));
+// resumeId = resume api link
```
Comment (DaleneyRuben) [inline src/components/Resume/ResumeDetail/ResumeDetail.jsx]: remove this comment

File: package.json (hunk 8 above)
Comment (alejandro-castellon) [reply package.json]: Sure, I left you a message in teams for this

File: src/components/Resume/ResumeDetail/ResumeDetail.jsx [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+
+  const edit = (event) => {
+    event.preventDefault();
+    setReadOnly(!readOnly);
+    if (readOnly) {
+      setShowEdit("Save");
+      alert("You can edit"); // change alert message
+    } else {
+      setShowEdit("Edit");
+      alert("Saved"); // change alert message
```
Comment (DaleneyRuben) [inline src/components/Resume/ResumeDetail/ResumeDetail.jsx]: remove alerts

File: src/components/Resume/ResumeDetail/ResumeDetail.jsx [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+    }
+  };
+  if (error) return <div>ERROR!!! Resume not found</div>;
+  return data ? (
```
Comment (DaleneyRuben) [inline src/components/Resume/ResumeDetail/ResumeDetail.jsx:38]: Confusing if there's no data it means is loading? 🤔 
maybe you can add a loading state in your hook to know if the data is loading or not

**Review by DaleneyRuben — CHANGES_REQUESTED** [review]

File: src/components/Resume/ResumeDetail/ResumeDetail.jsx [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+  root: {
+    margin: 30,
+    width: 500,
+  },
+}));
+
+function Resume({ resumeId }) {
+  const classes = useStyles();
+  const { data, error } = useFetch(resumeId);
+  console.log(data);
```
Comment (DaleneyRuben) [inline src/components/Resume/ResumeDetail/ResumeDetail.jsx]: Remove this console.log after that I'll give you the approval 🎉

**Review by DaleneyRuben — CHANGES_REQUESTED** [review]

**Review by DaleneyRuben — APPROVED** [review]
LGTM 👍🏻

**Review by jorge-lema-jalafund — APPROVED** [review]
Approved

**Pull request merged**